### ISSUE TITLE
PROTOCOL_ERROR after upgrading gRPC and Bigtable libraries

### ISSUE THREAD

**Issue by garye** [grpc/grpc-go#1247]
The issue was reported here: https://github.com/GoogleCloudPlatform/google-cloud-go/issues/628

The Bigtable libraries adds a string to the metadata but otherwise doesn't do much with contexts or metadata. Any ideas what could be going on?

**Comment by bbassingthwaite-va** [comment]
I've tried explicitly to do: ```metadata.NewOutgoingContext(ctx, nil)``` with no success either. Or would you expect this not to work?

**Comment by menghanl** [comment]
Do we know what the string added to metadata by Bigtable is? Does it contain special characters (like newline)?

**Comment by dfawley** [comment]
That would clear the outgoing metadata.  To add an entry to the outgoing metadata, you should do something like this:

```go
md, _ := metadata.FromOutgoingContext(ctx)
md = metadata.Join(md, metadata.New(map[string]string{"key": "value"}))
ctx = metadata.NewOutgoingContext(ctx, md)
```

**Comment by garye** [comment]
@menghanl We're adding a string without special characters or newlines, and we leverage this code:
https://github.com/GoogleCloudPlatform/google-cloud-go/blob/master/bigtable/bigtable.go#L779

Looks like @dfawley has a PR to fix this for us:
https://github.com/GoogleCloudPlatform/google-cloud-go/pull/627/files

What's the status of that?

**Comment by bbassingthwaite-va** [comment]
So if that is the case: Why does `context.Background()` work but not `metadata.NewOutgoingContext(ctx, nil)` ?

**Comment by dfawley** [comment]
It sounds like it's something else in the context, not the metadata, then...

Can you log the context?  It may also be helpful if you could provide a reproducible test case so we can dig into it.

**Comment by bbassingthwaite-va** [comment]
I haven't been able to get a test to reproduce this, there are a lot of layers in trying to reproduce this. Here is the context with both incoming and outgoing metadata being cleared before going to big table.

```golang
NewContext: (*context.valueCtx)(0xc423049290)(context.Background.WithDeadline(2017-05-1819: 26: 44.106173694+0000UTC[
  9.996859088s
]).WithValue(peer.peerKey{
  
},
&peer.Peer{
  Addr: (*net.TCPAddr)(0xc4229967e0),
  AuthInfo: credentials.TLSInfo{
    State: tls.ConnectionState{
      Version: 0x303,
      HandshakeComplete: true,
      DidResume: false,
      CipherSuite: 0xc02f,
      NegotiatedProtocol: "h2",
      NegotiatedProtocolIsMutual: true,
      ServerName: "<REDACTED>",
      PeerCertificates: [
        
      ]*x509.Certificate(nil),
      VerifiedChains: [
        
      ][
        
      ]*x509.Certificate(nil),
      SignedCertificateTimestamps: [
        
      ][
        
      ]uint8(nil),
      OCSPResponse: [
        
      ]uint8(nil),
      TLSUnique: [
        
      ]uint8{
        <REDACTED>
      }
    }
  }
}).WithValue(transport.streamKey{
  
},
<stream: 0xc4226df000,
/<Redacted>>).WithValue(metadata.mdIncomingKey{
  
},
metadata.MD{
  "user-agent": [
    
  ]string{
    "grpc-go/1.4.0-dev"
  },
  "authorization": [
    
  ]string{
    "Bearer <REDACTED>"
  },
  ":authority": [
    
  ]string{
    "<REDACTED>"
  }
}).WithValue(metadata.mdIncomingKey{
  
},
metadata.MD(nil)).WithValue(metadata.mdOutgoingKey{
  
},
metadata.MD(nil)))
```

**Comment by bbassingthwaite-va** [comment]
So I had tried clearing both incoming and outgoing, but never both. This line is what ended up working for me.  No idea why this works, but it does:
```golang
newCtx := metadata.NewOutgoingContext(
    metadata.NewIncomingContext(ctx, nil),
    nil,
)```

**Comment by dfawley** [comment]
Almost definitely what's happening here is the client interceptor that adds its metadata is reading your _incoming_ metadata (because it is using `metadata.FromContext`), appending its value to the metadata there, then using that as the outgoing metadata.  This is the main root cause, and that should be fixed by GoogleCloudPlatform/google-cloud-go#627.

In addition, it seems there's characters in the incoming metadata that isn't allowed in text-only outgoing metadata, and that's why we're getting the error we're getting.  It would be interesting to know the nature of the redacted authorization/authority values (e.g. are there characters in there outside the range of the base64 encoding set?).  Also, you said earlier that you have clients that are running both the latest grpc version and older versions, and only the older versions fail.  Can you compare the incoming metadata at the server from those two clients and see if you can discern a difference between them?

**Comment by dfawley** [comment]
The fix to the bigtable client library was merged, so this should be resolved now.
